telemetry/otel.go

```diff
@@ -0,0 +1,234 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package telemetry initializes the OpenTelemetry SDK for library users. It
+// primarily serves as a convenience package for setting up OpenTelemetry SDK
+// for GCP applications with some useful settings.
+//
+// Once initialized, users get automatic instrumentation of metrics and traces
+// and can start writing custom metrics and configuring custom traces.
+//
+// To initialize ... somewhere in your service setup code ...
+//
+//	shutdown, err := telemetry.Start(ctx)
+//	defer shutdown()
+//
+// Metrics and traces will then be exported to an OpenTelemetry collector endpoint
+// with default address set to localhost:4317.
+//
+// For testing before setting up and exporting to a collector, use
+// telemetry.StartWithStdoutExport instead.
+//
+// Much of this code is inspired from https://opentelemetry.io/docs/languages/go/getting-started/#initialize-the-opentelemetry-sdk
+package telemetry
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/contrib/detectors/gcp"
+	"go.opentelemetry.io/contrib/instrumentation/runtime"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/exporters/stdout/stdoutmetric"
+	"go.opentelemetry.io/otel/exporters/stdout/stdouttrace"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+	"go.opentelemetry.io/otel/sdk/resource"
+	"go.opentelemetry.io/otel/sdk/trace"
+)
+
+var defaultStartOnce sync.Once
+
+const (
+	defaultOtelAgentAddr = "0.0.0.0:4317"
+)
+
+type (
+	meterProviderFunc  func(context.Context, *resource.Resource) (*metric.MeterProvider, error)
+	tracerProviderFunc func(context.Context, *resource.Resource) (*trace.TracerProvider, error)
+)
+
+// Start initializes automatic instrumentation for exporting OTLP metrics and traces
+// to an OpenTelemetry collector agent.
+func Start(ctx context.Context) (shutdown func(context.Context) error, err error) {
+	defaultStartOnce.Do(func() {
+		shutdown, err = start(ctx, newOTLPTraceProvider, newOTLPMeterProvider)
+	})
+	return
+}
+
+// StartWithStdoutExport initializes automatic instrumentation for exporting OTLP
+// metrics and traces to stdout (e.g. Cloud Logging if using Cloud Run).
+//
+// This is intended for testing and debugging metrics and traces before exporting
+// them to an OpenTelemetry collector agent.
+func StartWithStdoutExport(ctx context.Context) (shutdown func(context.Context) error, err error) {
```

Comment:
Wonder if this would be better served with the options pattern

Reply:
Done.